Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs

```diff
@@ -327,38 +329,48 @@ private async Task<bool> RestoreCheckpointIncrementalAsync(OperationContext cont
             // Parse the checkpoint info for the checkpoint being restored
             var newCheckpointInfo = ParseCheckpointInfo(checkpointFile);
 
-            foreach (var entry in newCheckpointInfo)
-            {
-                var relativePath = entry.Key;
-                var storageId = entry.Value;
-
-                var incrementalCheckpointFile = _incrementalCheckpointDirectory / relativePath;
-
-                if (_incrementalCheckpointInfo.TryGetValue(relativePath, out var fileStorageId)
-                    && storageId == fileStorageId
-                    && _database.IsImmutable(incrementalCheckpointFile)
-                    && _fileSystem.FileExists(incrementalCheckpointFile))
-                {
-                    // File is already present in the incremental checkpoint directory, no need to download it
-                    await _storage.TouchBlobAsync(context, incrementalCheckpointFile, storageId, isUploader: false).ThrowIfFailure();
-                    Counters[ContentLocationStoreCounters.IncrementalCheckpointFilesDownloadSkipped].Increment();
-                }
-                else
-                {
-                    // File is missing, different, or mutable so download it and update it in the incremental checkpoint
-                    _fileSystem.DeleteFile(incrementalCheckpointFile);
-                    await _storage.TryGetFileAsync(context, storageId, incrementalCheckpointFile).ThrowIfFailure();
-                    Counters[ContentLocationStoreCounters.IncrementalCheckpointFilesDownloaded].Increment();
-                }
 
-                // Move the file from the incremental checkpoint into the extraction directory for loading by the database
-                await HardlinkWithFallBackAsync(context, incrementalCheckpointFile, checkpointTargetDirectory / relativePath);
+            foreach (var (key, value) in newCheckpointInfo)
+            {
+                await RestoreFileAsync(context, checkpointTargetDirectory, key, value).ThrowIfFailure();
             }
-
+            
             // Finalize by adding the incremental checkpoint info file to the local incremental checkpoint directory
             await HardlinkWithFallBackAsync(context, checkpointFile, _incrementalCheckpointInfoFile);
             UpdateIncrementalCheckpointInfo(newCheckpointInfo);
             return true;
         }
+
+        private Task<BoolResult> RestoreFileAsync(OperationContext context, AbsolutePath checkpointTargetDirectory, string relativePath, string storageId, bool forceUseExistingFiles = false)

```

Comment:
Maybe a simple comment explaining the motivation of forceUseExistingFiles? Specially since it is not used anywhere, as far as I can tell

Reply:
It was used in one of my experiements. I removed it completely.

---
In reply to: [275919167](https://github.com/Microsoft/BuildXL/pull/150#discussion_r275919167) [](ancestors = 275919167)